### DISCUSSION TITLE
Volume not adjustable - Max98257 / ESP32

### DISCUSSION THREAD

**Discussion by DonRonito** [pschatzmann/arduino-audio-tools#814] [Q&A] [answered]
I'm currently running the following code, but can't seem to get the volume adjusted. No matter the value the speaker is playing 100%. What am I missing? :)

`/**
 * @file streams-url_mp3-i2s.ino
 * @author Phil Schatzmann
 * @brief decode MP3 stream from url and output it on I2S
 * @version 0.1
 * @date 2021-96-25
 * 
 * @copyright Copyright (c) 2021
 */

// install https://github.com/pschatzmann/arduino-libhelix.git

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"

URLStream url("abcd","1234");
I2SStream i2s; // final output of decoded stream
VolumeStream volume(i2s);
EncodedAudioStream dec(&i2s, new MP3DecoderHelix()); // Decoding stream
StreamCopy copier(dec, url); // copy url to decoder

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  // you could define e.g your pins and change other settings
  config.pin_ws = 21;
  config.pin_bck = 18;
  config.pin_data = 19;
  //config.mode = I2S_STD_FORMAT;
  i2s.begin(config);

  // set initial volume
  volume.begin(config); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.3);

  // setup I2S based on sampling rate provided by decoder
  dec.setNotifyAudioChange(i2s);
  dec.begin();

// mp3 radio
  url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");

}

void loop(){
  copier.copy();
}
`

**Comment by pschatzmann** [comment, marked as answer]
You send the output directly to i2s instead of to volume, so volume is not used anywhere!

**Reply by DonRonito** [reply]
Thank you. I edited to the following and it's now working :).

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"

URLStream url("abcd","1234");
I2SStream i2s; // final output of decoded stream
VolumeStream volume(i2s);
EncodedAudioStream dec(&volume, new MP3DecoderHelix()); // Decoding stream
StreamCopy copier(dec, url); // copy url to decoder

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup i2s
  auto config = i2s.defaultConfig(TX_MODE);
  config.pin_ws = 21;
  config.pin_bck = 18;
  config.pin_data = 19;
  i2s.begin(config);

  // setup I2S based on sampling rate provided by decoder
  dec.setNotifyAudioChange(i2s);
  dec.begin();

  // set initial volume
  volume.begin(config); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.2);

  // mp3 radio
  url.begin("http://stream.srg-ssr.ch/m/rsj/mp3_128","audio/mp3");
}

void loop(){
  copier.copy();
}
